hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CreatePipelineCommandHandler.java

```diff
@@ -0,0 +1,226 @@
+        cmdStatus.setStatus(cmdExecuted);
+        ((CreatePipelineCommandStatus)cmdStatus)
+            .setCreatePipelineAck(createPipelineACK);
+      };
+      updateCommandStatus(context, command, statusUpdater, LOG);
+      long endTime = Time.monotonicNow();
+      totalTime += endTime - startTime;
+    }
+  }
+
+  /**
+   * Returns the command type that this command handler handles.
+   *
+   * @return Type
+   */
+  @Override
+  public SCMCommandProto.Type getCommandType() {
+    return SCMCommandProto.Type.createPipelineCommand;
+  }
+
+  /**
+   * Returns number of times this handler has been invoked.
+   *
+   * @return int
+   */
+  @Override
+  public int getInvocationCount() {
+    return invocationCount;
+  }
+
+  /**
+   * Returns the average time this function takes to run.
+   *
+   * @return long
+   */
+  @Override
+  public long getAverageRunTime() {
+    if (invocationCount > 0) {
+      return totalTime / invocationCount;
+    }
+    return 0;
+  }
+
+  /**
+   * Sends ratis command to create pipeline on this datanode.
+   *
+   * @param dn       - this datanode
+   * @param pipelineId   - pipeline ID
+   * @param peers        - datanodes of the pipeline
+   * @param ozoneConf  - Ozone configuration
+   * @throws IOException
+   */
+  private void createPipeline(DatanodeDetails dn, PipelineID pipelineId,
+      Collection<DatanodeDetails> peers, Configuration ozoneConf)
+      throws IOException {
+    final RaftGroup group = RatisHelper.newRaftGroup(
+        RaftGroupId.valueOf(pipelineId.getId()), peers);
+    LOG.debug("creating pipeline:#{} with {}", pipelineId, group);
+
+    final String rpcType = ozoneConf
+        .get(ScmConfigKeys.DFS_CONTAINER_RATIS_RPC_TYPE_KEY,
+            ScmConfigKeys.DFS_CONTAINER_RATIS_RPC_TYPE_DEFAULT);
+    final RetryPolicy retryPolicy = RatisHelper.createRetryPolicy(ozoneConf);
+    final List< IOException > exceptions =
+        Collections.synchronizedList(new ArrayList<>());
+    final int maxOutstandingRequests = ozoneConf.getInt(
+        ScmConfigKeys.SCM_CONTAINER_CLIENT_MAX_OUTSTANDING_REQUESTS,
+        ScmConfigKeys.SCM_CONTAINER_CLIENT_MAX_OUTSTANDING_REQUESTS_DEFAULT);
+    final GrpcTlsConfig tlsConfig = RatisHelper.createTlsClientConfig(new
+        SecurityConfig(ozoneConf));
+    final TimeDuration requestTimeout =
+        RatisHelper.getClientRequestTimeout(ozoneConf);
+    try {
+      final RaftPeer peer = RatisHelper.toRaftPeer(dn);
+      try (RaftClient client = RatisHelper
+          .newRaftClient(SupportedRpcType.valueOfIgnoreCase(rpcType), peer,
+              retryPolicy, maxOutstandingRequests, tlsConfig,
+              requestTimeout)) {
+
+        RaftClientReply reply = client.groupAdd(group, peer.getId());
+        if (reply == null || !reply.isSuccess()) {
+          String msg = "Pipeline initialization failed for pipeline:"
+              + pipelineId.getId() + " node:" + peer.getId();
+          LOG.error(msg);
```

Comment:
LOG.error can be removed as the caller has another LOG error for the same exception threw.

Reply:
OK